web/components/login/Login.js

```diff
@@ -8,53 +8,62 @@ import { connect } from 'react-redux';
 class Login extends React.Component {
   constructor(props) {
     super(props);
+    const { auth: { isLoaded, isEmpty } } = this.props;
 
     this.state = {
       email: '',
       password: '',
-      loading: !this.props.auth.isLoaded,
-      loggedIn: !this.props.auth.isEmpty,
+      loading: !isLoaded,
+      loggedIn: !isEmpty,
     };
 
-    this.onEmailChange = (event) => {
-      this.setState({ email: event.target.value });
-    };
+    this.onEmailChange = event => this.setState({ email: event.target.value });
+    this.onPasswordChange = event => this.setState({ password: event.target.value });
 
-    this.onPasswordChange = (event) => {
-      this.setState({ password: event.target.value });
-    };
+    this.login = this.login.bind(this);
+    this.loginView = this.loginView.bind(this);
+    this.signedInView = this.signedInView.bind(this);
```

Comment:
What's the reason behind using `.bind(this)` here? Seems like signedInView is only used inside this class and thus you don't have any dynamic scope issues with `this`.

`login` needs to be bound since it's called from external functions, but not the internal view methods. `bind` is hideously slow and should be avoided whenever possible (90% slower).

https://jsperf.com/function-calls-direct-vs-apply-vs-call-vs-bind/14

Reply:
wow TIL I'll have a closer look at this, thanks!

Reply:
Where do you see `login()` being called externally? I think none of these functions need a `this` binding, since they're only called within this component, but maybe im misunderstanding

Reply:
Might also consider using https://github.com/andreypopp/autobind-decorator instead. Makes the code more readable.

Could also try https://babeljs.io/docs/en/babel-plugin-transform-class-properties/ which lets you do `login = () => {}` at the class level.

Reply:
put up #14 to explore the babel plugin